### PR TITLE
VDS-790: Load product parts of the configurable product with him

### PR DESCRIPTION
**Pull request by trueboroda**

### CODE REVIEW

File: VirtoCommerce.Storefront/Domain/Catalog/DemoCatalogService.cs [hunk 1]
```diff
@@ -17,124 +21,70 @@ namespace VirtoCommerce.Storefront.Domain.Catalog
...
-            memoryProductPart.Items = memoryProducts;
-            memoryProductPart.SelectedItemId = memoryProducts.FirstOrDefault()?.Id;
+            await base.LoadProductDependencies(products, responseGroup, workContext);
 
-            var gpuProductPart = TryGetProductPartByCategoryId(categories["GPU"]);
-            gpuProductPart.Items = gpuProducts;
-            gpuProductPart.SelectedItemId = gpuProducts.FirstOrDefault()?.Id;
-
-            var parts = new[]
+            if (!products.IsNullOrEmpty())
```
Comment (Vectorfield4) [inline VirtoCommerce.Storefront/Domain/Catalog/DemoCatalogService.cs:52]: Is `products=null` are viable value for this place? if not we can remove this condition.

File: VirtoCommerce.Storefront/Domain/Catalog/DemoCatalogService.cs [hunk 2]
```diff
@@ -17,124 +21,70 @@ namespace VirtoCommerce.Storefront.Domain.Catalog
...
-                {
-                    "f49f255f-512b-4172-ba3c-cfaa8995c159",
-                    new ProductPart
-                    {
-                        Name = "Graphics",
-                        Image = new Image { Url = "https://raw.githubusercontent.com/VirtoCommerce/vc-demo-theme-b2b/dev/assets/images/mock/graphics.svg" },
-                    }
-                },
-            };
+                cacheEntry.AddExpirationToken(CatalogCacheRegion.CreateChangeToken());
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront/Domain/Catalog/DemoCatalogService.cs:71]: Add api changes token

File: VirtoCommerce.Storefront/Domain/Catalog/Demo/CatalogConverter.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+using demoCatalogDto = VirtoCommerce.Storefront.AutoRestClients.DemoSolutionFeaturesModuleModuleApi.Models;
+
+
+namespace VirtoCommerce.Storefront.Domain
+{
+    public static partial class CatalogConverter
+    {
+        public static ProductPart ToProductPart(this demoCatalogDto.DemoProductPart productPartDto, string currentLanguage)
+        {
+            var result = new ProductPart()
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront/Domain/Catalog/Demo/CatalogConverter.cs]: remove ()

**Review by pushnitsa — CHANGES_REQUESTED** [review]

File: VirtoCommerce.Storefront/Domain/Catalog/DemoCatalogService.cs [hunk 4]
```diff
@@ -17,124 +21,70 @@ namespace VirtoCommerce.Storefront.Domain.Catalog
...
-                motherboardProductPart,
-                cpuProductPart,
-                memoryProductPart,
-                gpuProductPart,
-            };
-
-            return parts.OrderBy(x => x.Name).ToArray();
+                foreach(var product in products)
+                {
+                    if(product.ProductType.EqualsInvariant("Configurable"))
```
Comment (Vectorfield4) [inline VirtoCommerce.Storefront/Domain/Catalog/DemoCatalogService.cs]: You can move this condition to line 52 to reduce cognitive complexity. 
``` csharp
// Load product parts to configurable products
foreach(var product in products.Where(x => x.ProductType.EqualsInvariant("Configurable")))
```

File: VirtoCommerce.Storefront/Domain/Catalog/DemoCatalogService.cs [hunk 5]
```diff
@@ -17,124 +21,70 @@ namespace VirtoCommerce.Storefront.Domain.Catalog
...
-                    "f49f255f-512b-4172-ba3c-cfaa8995c159",
-                    new ProductPart
-                    {
-                        Name = "Graphics",
-                        Image = new Image { Url = "https://raw.githubusercontent.com/VirtoCommerce/vc-demo-theme-b2b/dev/assets/images/mock/graphics.svg" },
-                    }
-                },
-            };
+                cacheEntry.AddExpirationToken(CatalogCacheRegion.CreateChangeToken());
+                var searchPartsResultDto = await _demoCatalogApi.SearchAsync(new DemoProductPartSearchCriteria() { ConfiguredProductId = productId, Take = 1000 });
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront/Domain/Catalog/DemoCatalogService.cs]: Take = int.maxValue

**Review by pushnitsa — CHANGES_REQUESTED** [review]

File: VirtoCommerce.Storefront/Domain/Catalog/DemoCatalogService.cs [hunk 6]
```diff
@@ -17,124 +21,70 @@ namespace VirtoCommerce.Storefront.Domain.Catalog
...
+            var allPartItems = await GetProductsAsync(itemsIds); // Potential recursion
 
-            return result.ContainsKey(categoryId) ? result[categoryId] : null;
+            var productParts = searchResult.Results.OrderBy(x => x.Priority).Select(x => {
+                var productPart = x.ToProductPart(workContext.CurrentLanguage.CultureName);
+                productPart.Items = x.PartItems.OrderBy(itemDto => itemDto.Priority)
+                                    .Select(itemDto =>  allPartItems.FirstOrDefault(p => p.Id.EqualsInvariant(itemDto.ItemId)))
+                                    .Where(p=>p != null).ToArray();
+                return productPart;
+                }).ToArray();
```
Comment (Vectorfield4) [inline VirtoCommerce.Storefront/Domain/Catalog/DemoCatalogService.cs]: Move select function to local function/method. It's hard to read.

**Review by Vectorfield4 — CHANGES_REQUESTED** [review]

File: VirtoCommerce.Storefront/Domain/Catalog/DemoCatalogService.cs (hunk 1 above)
Comment (trueboroda) [reply VirtoCommerce.Storefront/Domain/Catalog/DemoCatalogService.cs:52]: Such behavior in the original method LoadProductDependencies.

Comment (trueboroda) [reply VirtoCommerce.Storefront/Domain/Catalog/DemoCatalogService.cs:52]: https://github.com/VirtoCommerce/vc-demo-storefront/blob/e870ffc2c91d468fe7155065f022ba3a082c8ee1/VirtoCommerce.Storefront/Domain/Catalog/CatalogService.cs#L213

**Review by Vectorfield4 — APPROVED** [review]

File: VirtoCommerce.Storefront.Model/Catalog/Demo/ProductTypes.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace VirtoCommerce.Storefront.Model.Catalog.Demo
+{
+    public static class ProductTypes
+    {
+        static public string Configurable { get; set; } = "Configurable";
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront.Model/Catalog/Demo/ProductTypes.cs]: replace property by constant

File: VirtoCommerce.Storefront.Model/Catalog/Specifications/ProductIsBuyableSpecification.cs [hunk 8]
```diff
@@ -10,7 +12,8 @@ public virtual bool IsSatisfiedBy(Product product)
             if (product == null)
                 throw new ArgumentNullException(nameof(product));
 
-            return product.IsActive && product.IsBuyable && product.Price.ListPrice.Amount > 0;
+            return ( product.IsActive && product.IsBuyable && product.Price.ListPrice.Amount > 0 && product.ProductType != ProductTypes.Configurable )
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront.Model/Catalog/Specifications/ProductIsBuyableSpecification.cs:15]: Add unit tests to this method

**Review by pushnitsa — CHANGES_REQUESTED** [review]

**Review by Vectorfield4 — APPROVED** [review]

**Review by pushnitsa — APPROVED** [review]

**Pull request merged**